@here/harp-geometry/lib/TechniqueProcessor.ts

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getPropertyValue } from "@here/harp-datasource-protocol";
+import { MapEnv } from "@here/harp-datasource-protocol/index-decoder";
+import { MapView } from "@here/harp-mapview";
+import * as THREE from "three";
+
+/**
+ * Base interface for all [[Technique]]-related configuration paramets.
+ */
+export interface TechniqueParams {
+    /**
+     * [[Technique]] identifier.
+     */
+    readonly name: string;
+    /**
+     * The render order of the objects created using this technique.
+     *
+     * If not specified in style file, [[StyleSetEvaluator]] will assign monotonically increasing
+     * values according to style position in file.
+     */
+    readonly renderOrder: number;
+}
+
+/**
+ * Base class that handles processing data accordingly to its assigned [[Technique]].
+ */
+export abstract class TechniqueProcessor {
```

Comment:
While i agree, that unification of geometry processing is needed, i will raise key problem from the beginning:

Separation of "decoder" code that lands in worker bundle and "main thread" code.

Class based design binds all this code `createBuffers` and `vertexShader` and finally `createObjects` into one piece of unbreakable code and we result with `MapView` (unnecessarily) and friends linked into decoders ... and vice versa - "createBuffers" linked into main thread bundle.

_If_ this is a concern, then unfortunately, `TechniqueProcessor` has to be more of like a "concept" how to structure code and maybe set of interfaces and implementations, but they cannot be one language entity (class, function, record), so they are really separable from link time perspective.

Reply:
I had a quick chat with Andres yesterday and we already identified that the technique processors have to be registered twice:
1. In MapView for the main thread to create objects
2. Somewhere else for the decoders to create buffers.
So you are absolutely right, this could be split in two classes, one for the main thread and one for the web worker.

Reply:
Splitting this into 2 classes will make it harder to maintain, since it's no longer a single class per technique. ``createBuffers``can be called from both the main thread and the decoding threads (as showcased in one of our examples), and ``createObjects`` should never be called from decoders, since it's aimed to be connected to MapView, as you guys stated.

You're not supposed to instantiate this classes (they're fully static classes, maybe we can change to singletons?), so even if the code is both on the main and decoder bundles, I think we should use these methods wisely (no need to link MapView to the decoders) and keep them in a single coherent place where we can understand the data flow from the fetched tiles to the three.js objects on screen.

Reply:
> Splitting this into 2 classes will make it harder to maintain, since it's no longer a single class per technique. 

Yes, it will be a little harder. Unfortunately, we having decoders makes us "client-server" architecture where client/server implementation shouldn't leak to each other - even if they are very tightly coupled.
We have to work with this "cost".

Now, with `SceneState` we're much better, because at least MapView doesn't leak so much, issue is still there.

What i'm trying to say, is that we should split this "buffer" creation and "3d objects" creation to separate classes language entities, preferably files, so "linker" (webpack/rollup) can link only needed functions in respective bundles.

Specifically, we don't want this in decoder bundle:
 * we don't want shaders in bundles
 * we don't want whole shaders, Object3d and Meshes, PerspectiveCamera and stuff in decoder bundlers

And we don't "buffer creation code"  like `createLines`, `Extuder",  `Outliner" in main bundle.

...
Yes, someone may say that JS code is cheap, but our bundles are already "500k" gzipped/minified or more. They shouldn't grow without a reason.

Reply:
> ~500k

Ok, i was too pessimistic, actual sizes are not so bad:
```
200K	./harp-decoders.min.js
56K	./harp-decoders.min.js.gz
116K	./harp.min.js.gz
460K	./harp.min.js
```